src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -1002,7 +1002,98 @@ class segment_meta_cstore::impl
         _write_buffer.erase(it);
     }
 
-    void insert(const segment_meta& m) {
+    bool can_be_inserted(const segment_meta& m) const {
+        // segment insertion is order dependant, so not all the sequence of
+        // insertions of the same set of segment_meta will result in the same
+        // end result. for example. given this manifest [0, 10][11, 20],
+        // inserting the segment [9, 12] is symptomatic of an underlying issue
+        // (no alignment whatsoever), but can't be rejected immediately. other
+        // edge cases can be rejected in the interest of keeping the manifest in
+        // a reasonable state: if the new segment is entirely inside an old
+        // segment it can be rejected. in this case [12, 20], [18,20] and [14,
+        // 17] should be rejected.
+        // in case of perfect overlap, the insertion can be done since it does
+        // not break the invariants, but the user of this class
+        // (partition_manifest) will apply further logic to restrict this case
```

Comment:
I'm not sure it's possible, but I'm trying to decompose this into other primitives that might be easier to reason about as a caller. Aside from reading this method, I'm having a hard time trying to articulate the conditions we can insert. I'm thinking generally something like:

```
bool can_be_inserted(sm) {
  if (sm.base_offset < base_offset) {
    return false
  }
  return has_perfect_overlap(sm) || segment_has_uncovered_offsets(sm);
}
```

Is that on the right track?

Reply:
can_be_inserted just deals with new segments that would be span offsets already covered by another segment in the cstore
so for example for a cstore with this segment [base_offset 10 committed_offset 20] any new segment [base_offset (11...20) committed_offset (11, 20)] would be rejected.

the problem stems from the fact that the merge operation requires the operation < to have a reasonable meaning for segment_meta. 

if we define operator<(left, right) -> return left.base_offset < right.base_offset, and we don't restrict the input, then a 
cstore [10, 20][11, 19] would be legal, and the call last_segment would return [11, 19], even though the appropriate result should be [10, 20]

in term of your function, i think it should be
```cpp
bool can_be_inserted(sm) {
  return has_perfect_overlap(sm) || segment_has_uncovered_offsets(sm);
}
```